File: rerun_cpp/src/rerun/datatypes/annotation_info_ext.cpp
#include <utility>
#include "annotation_info.hpp"

// Uncomment for better auto-complete while editing the extension.
// #define EDIT_EXTENSION

namespace rerun {
    namespace datatypes {

#ifdef EDIT_EXTENSION
        struct AnnotationInfoExt {
            uint16_t id;
            std::optional<datatypes::Utf8> label;
            std::optional<datatypes::Rgba32> color;

#define AnnotationInfo AnnotationInfoExt

            // <CODEGEN_COPY_TO_HEADER>

            AnnotationInfo(
                uint16_t _id, std::optional<std::string> _label = std::nullopt,
                std::optional<datatypes::Rgba32> _color = std::nullopt
            )
                : id(_id), label(std::move(_label)), color(_color) {}

            AnnotationInfo(uint16_t _id, datatypes::Rgba32 _color)
                : id(_id), label(std::nullopt), color(_color) {}

            // </CODEGEN_COPY_TO_HEADER>
        };
#endif

    } // namespace datatypes
} // namespace rerun

File: crates/store/re_sdk_types/src/encodings/annotation_info_ext.rs
use super::{AnnotationInfo, Rgba32};

impl From<u16> for AnnotationInfo {
    fn from(id: u16) -> Self {
        Self {
            id,
            label: None,
            color: None,
        }
    }
}

impl From<(u16, &str)> for AnnotationInfo {
    fn from((id, label): (u16, &str)) -> Self {
        Self {
            id,
            label: Some(label.into()),
            color: None,
        }
    }
}

impl From<(u16, &str, Rgba32)> for AnnotationInfo {
    fn from((id, label, color): (u16, &str, Rgba32)) -> Self {
        Self {
            id,
            label: Some(label.into()),
            color: Some(color),
        }
    }
}
